flink-runtime/src/main/java/org/apache/flink/runtime/jobmaster/slotpool/DefaultDeclarativeSlotPool.java

```diff
@@ -354,30 +369,49 @@ public void registerNewSlotsListener(NewSlotsListener newSlotsListener) {
 
     @Override
     public ResourceCounter releaseSlots(ResourceID owner, Exception cause) {
-        final Collection<AllocatedSlot> removedSlots = slotPool.removeSlots(owner);
+        final AllocatedSlotPool.AllocatedSlotsAndReservationStatus removedSlots =
+                slotPool.removeSlots(owner);
 
-        ResourceCounter previouslyFulfilledRequirements = getFulfilledRequirements(removedSlots);
-
-        releasePayload(removedSlots, cause);
-        releaseSlots(removedSlots, cause);
+        final Collection<AllocatedSlot> slotsToFree = new ArrayList<>();
+        for (AllocatedSlot removedSlot : removedSlots.getAllocatedSlots()) {
+            if (!removedSlots.wasFree(removedSlot)) {
+                slotsToFree.add(removedSlot);
+            }
+        }
 
-        return previouslyFulfilledRequirements;
+        return freeAndReleaseSlots(slotsToFree, removedSlots.getAllocatedSlots(), cause);
     }
 
     @Override
     public ResourceCounter releaseSlot(AllocationID allocationId, Exception cause) {
```

Comment:
The behavior of these 2 methods should be identical, so some common parts were extracted.